hyperspy/_signals/signal2d.py

```diff
@@ -605,3 +608,433 @@ def add_ramp(self, ramp_x, ramp_y, offset=0):
+    """
+    Method to locate positive peaks in an image based on gradient
+    thresholding and subsequent refinement within masked regions.
+    Parameters
+    ----------
+    z : ndarray
+        Matrix of image intensities.
+    grad_threshold : float
+        The minimum gradient required to begin a peak search.
+    window_size : int
+        The size of the square window within which a peak search is
+        conducted. If odd, will round down to even.
+    distance_cutoff : float
+        The maximum distance a peak may be from the initial
+        high-gradient point.
+    Returns
+    -------
+    peaks : numpy.ndarray
+        (n_peaks, 2)
+        Peak pixel coordinates.
+    Notes
+    -----
+    Implemented as described in Zaefferer "New developments of computer-aided
+    crystallographic analysis in transmission electron microscopy" J. Ap. Cryst.
+    This version by Ben Martineau (2016)
+    """
+
+    def box(x, y, window_size, x_max, y_max):
+        """Produces a list of coordinates in the box about (x, y)."""
+        a = int(window_size / 2)
+        x_min = max(0, x - a)
+        x_max = min(x_max, x + a)
+        y_min = max(0, y - a)
+        y_max = min(y_max, y + a)
+        return np.array(
+            np.meshgrid(range(x_min, x_max), range(y_min, y_max))).reshape(
+            2,
+            -1).T
+
+    def get_max(image, box):
+        """Finds the coordinates of the maximum of 'image' in 'box'."""
+        vals = image[box[:, 0], box[:, 1]]
+        max_position = box[np.argmax(vals)]
+        return max_position
+
+    def distance(x, y):
+        """Calculates the distance between two points."""
+        v = x - y
+        return np.sqrt(np.sum(np.square(v)))
+
+    def gradient(image):
+        """Calculates the square of the 2-d partial gradient.
+
+        Parameters
+        ----------
+        image : ndarray
+
+        Returns
+        -------
+        ndarray
+
+        """
+        gradient_of_image = np.gradient(image)
+        gradient_of_image = gradient_of_image[0] ** 2 + gradient_of_image[
+                                                            1] ** 2
+        return gradient_of_image
+
+    # Generate an ordered list of matrix coordinates.
+    if len(z.shape) != 2:
+        raise ValueError("'z' should be a 2-d image matrix.")
+    z = z / np.max(z)
+    coordinates = np.indices(z.data.shape).reshape(2, -1).T
+    # Calculate the gradient at every point.
+    image_gradient = gradient(z)
+    # Boolean matrix of high-gradient points.
+    gradient_is_above_threshold = image_gradient >= grad_threshold
+    peaks = []
+    for coordinate in coordinates[gradient_is_above_threshold.flatten()]:
+        # Iterate over coordinates where the gradient is high enough.
+        b = box(coordinate[0], coordinate[1], window_size, z.shape[0],
+                z.shape[1])
+        p_old = np.array([0, 0])
+        p_new = get_max(z, b)
+        while np.all(p_old != p_new):
```

Comment:
This will continue searching for the peak even when we are too far away to save it later. I suggest combining the `while` loop with the `if` statement that follows it (likely using `break`) to stop the loop once it wanders too far from the start